### ISSUE TITLE
upon IBM Cloud timeout: Cannot read property 'statusCode' of undefined

### ISSUE THREAD

**Issue by shevkoplyas** [pmpkin/softlayer-node#14]
hi Michael!

Thank you for your great library!

On the bad day IBM was simply not responding to the posted orders to create new ubuntu boxes. Mostly it was in form of ~5 minutes timeout, once was simple connection reset about 15 seconds after request was posted. The problem with this library is that it has f-n "_handleResponse()" defined in this file:
/usr/src/app/node_modules/softlayer-node/lib/http-requests.js

, which goes like this:
```
var _handleResponse = function(callback) {
    return function(err, res, body) {
        if (res.statusCode >= 400 && res.statusCode < 600 || res.statusCode < 10) {
```
and upon those connection resets or 5min timeouts the "res" is undefined, thus whole provisioning app crashing with error: "Cannot read property 'statusCode' of undefined"

Also related question: why the .catch() block in the call example below does not catch the above error?  How to properly handle this case?
Here's how I call it:

```
let cloud_provider_response_raw = await softlayer_client
        .auth(this.config.softlayer_username, this.config.softlayer_api_key)
        .path("Product_Order", "placeOrder")
        .parameters(ibm_order)
        .post()
        .then() { /* we're happy block goes here */ }
        .catch(error => { /* we're sad block goes here */ }
```

so, somehow the .catch()  does not catch this error!

please suggest.
thanks,
Dmitry Shevkoplyas

**Comment by pmpkin** [comment]
Hi Dmitry,

I have to admit, that it's been a long time, since I had a look at this package. You're right, there should be a check if the response is null or undefined. But you can definitely catch this error.

I think your code sample is wrong. The post() returns a Promise. You can either await this promise, or you can add the then().catch() block, not both. If you do it with async/await, you have to put the statement in a catch block, to catch exceptions. This is how I would do it:

```javascript
let cloud_provider_response_raw;
try {
  cloud_provider_response_raw = await softlayer_client
        .auth(this.config.softlayer_username, this.config.softlayer_api_key)
        .path("Product_Order", "placeOrder")
        .parameters(ibm_order)
        .post();
} catch (e) {
// handle error here
}
```

If you want to do it with .then() and .catch():

```javascript
softlayer_client
        .auth(this.config.softlayer_username, this.config.softlayer_api_key)
        .path("Product_Order", "placeOrder")
        .parameters(ibm_order)
        .post()
        .then(cloud_provider_response_raw =>  /* we're happy block goes here */  ) 
        .catch(error => { /* we're sad block goes here */ }
```

I hope this helps :-)

**Comment by shevkoplyas** [comment]
Hi Michael,

Thank you for quick response!
I've tried to remove .then().catch()  and wrapped .post() into try/wait
block in hope to catch this exception, but it slips away anyway!-)
On the server I'm working on the whole thing is dockarized and must be
deployed via Jenkins build after each change is pushed into the private
github repo, so to ease the debug I've isolated/simplified that whole thing
into 1-pager JS file, which I pasted on the "pastebin" (had to register
there 5 minutes ago after 10+ years of passively using it due to today's
occasion;)

so here's the "hello world" that reproduces it if I may:
https://pastebin.com/kg0es83N

I'm trying to make main_with_await() to catch "Cannot read property
'statusCode' of undefined" error, but it seems is fired away bypassing all
my hopes : )

thanks,
Dmitry Shevkoplyas

**Comment by pmpkin** [comment]
Thanks for reproducing the error in this pastebin. I think, the easiest way to fix this, is to first check for an error. Now it works like this (http-requests.js, line 54)

````javascript
if (res.statusCode >= 400 && res.statusCode < 600 || res.statusCode < 10) {
 
} else if (err) 

}
````
But it should be:
````javascript
if (err) {
  callback(err);  
} else if (res.statusCode >= 400 && res.statusCode < 600 || res.statusCode < 10) 
  
}
````
If there's a timeout or another network error, the error should be correctly handled. I will look into this monday morning.

**Comment by shevkoplyas** [comment]
Hi Michael,

Yes, your suggested change would fix the problem and won't brake existing
functionality.
I wonder when "npm install" on my empty ubuntu box will be able to get your
fixed version? (in other words - when this change will propagate into
official ntp repository?)

Also, as JavaScript learner I'm still curious how one can catch that
uncatchable exception in my pastebin example?  I tried to add try/catch
literally everywhere I could (except inside your library code where the
actual error occur) and none of my try/catch blocks were able to catch it!

cheers,
Dmitry

**Comment by pmpkin** [comment]
Should be fixed with 0.2.2
